src/Disks/DiskCache.cpp

```diff
@@ -0,0 +1,1026 @@
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesCached, size);
+                    break;
+                case DiskCachePolicy::CachePart::CachePartType::ABSENT:
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesRemote, size);
+                    ProfileEvents::increment(ProfileEvents::S3ReadBytes, size);
+                    break;
+                case DiskCachePolicy::CachePart::CachePartType::ABSENT_NO_CACHE:
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesRemote, size);
+                    ProfileEvents::increment(ProfileEvents::DiskCacheBytesRemoteSkipCache, size);
+                    ProfileEvents::increment(ProfileEvents::S3ReadBytes, size);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        return res;
+    }
+
+    void nextIfAtEnd() override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    bool eof() override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    void ignore(size_t) override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    size_t tryIgnore(size_t) override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    void ignoreAll() override
+    {
+        throw Exception("Method not implemented", ErrorCodes::NOT_IMPLEMENTED);
+    }
+
+    bool peek(char & c) override
+    {
+        return from->peek(c);
+    }
+
+    bool read(char & c) override
+    {
+        return from->read(c);
+    }
+
+    void readStrict(char & c) override
+    {
+        from->readStrict(c);
+    }
+
+    size_t read(char * to, size_t n) override
+    {
+        return from->read(to, n);
+    }
+
+    void readStrict(char * to, size_t n) override
+    {
+        from->readStrict(to, n);
+    }
+
+    void prefetch() override
+    {
+        from->prefetch();
+    }
+
+private:
+    void ensureFrom()
+    {
+        int retries = 3;
+
+        while (!from)
+        {
+            size_t max_size = end_offset ? end_offset - current_offset : 0;
+            DiskCachePolicy::CachePart part(current_offset, max_size);
+            if (retries > 0)
+                part = cache_policy->find(cache_key, current_offset, max_size);
+            else
+            {   // Max retry attempts reached
+                // Came here when part present in index but absent on disk, several times
+                LOG_WARNING(log, "ensureFrom attempts max retries for key {}", cache_key);
+                part.type = DiskCachePolicy::CachePart::CachePartType::ABSENT_NO_CACHE;
+            }
+
+            if (part.type == DiskCachePolicy::CachePart::CachePartType::EMPTY)
+                return;
+
+            if (part.type == DiskCachePolicy::CachePart::CachePartType::CACHED)
```

Comment:
hm, looks like you'll need to support methods `setReadUntilPosition()` and `setReadUntilEnd()` (overriding from `ReadBuffer`) for buffer which reads from local cache.
Because `AsynchronousReadIndirectBufferFromRemoteFS` assumes that reading is always done with ranges (for merge tree family tables) and counts on the size of `working_buffer` to correspond to what methods `setReadUntilPosition()` and `setReadUntilEnd()` have said the buffer to read. So currently reading from local filesystem always reads up to `buffer_size` bytes into `working_buffer`, which later ends up in `AsynchronousReadIndirectBufferFromRemoteFS` and as a result everything gets completely broken.

Reply:
Yes, setReadUntilPosition() and setReadUntilEnd() are required. I have added them.